### ISSUE TITLE
Passing path names to Filesystem instead of inode numbers

### ISSUE THREAD

**Issue by mohsen3** [zargony/rust-fuse#75]
In the todo section of the README file, it is mentioned:
> An additional more high level API would be nice. It should provide pathnames instead inode numbers ...

This is what I need. I wonder if there has been any progress on this? I changed some of the functions such as `getattr` and `readdir` in the Filesystem trait to take `Path` as a parameter. But later, I noticed that the path only contains the current directory name (rather than the full path).

Is there any way to receive the full path in `Filesystem`? I can definitly make a higher level abstraction that keeps track of `inodes <-> paths` in a dictionary and passes the full path to the functions instead. Is there any easier/more efficient way to get the full path?

**Comment by zargony** [comment]
The idea is not to modify the current API to take paths, but to have an additional API.
Using paths is convenient, but it adds the overhead of lookups or conversion functions since the kernel internally always works with inodes. So, the existing Filesystem trait is supposed to stay like this (libfuse has fuse_lowlevel_ops) for a minimal-overhead interface.
An API with paths would internally again use the API with inodes and "just" needs to convert between paths and inodes. This shouldn't be too hard and could be done with lookup tables (HashMap) or by using path hashes as inode numbers. But unfortunately, nobody has taken the time to design it yet.

**Comment by roblabla** [comment]
I came accross https://github.com/anowell/netfuse just now, which seems to be good enough for most basic operations (it's missing lots of ops though).

**Comment by wfraser** [comment]
I'm also working on https://github.com/wfraser/fuse-mt which sits on top of rust-fuse and does inode->path translation, and adds multi-threading for some I/O operations. It's a work-in-progress that I haven't had much time to work on lately, though.

**Comment by droundy** [comment]
I'll add mention that a path name based API had the disadvantage of making it impossible to create a posix compliant (aka correct) file system. Unless you do not, for instance allow file deletion.  So it would be ideal to make this clear in the documentation for any such API. 

A path name based API is not a safe and easy API, but rather an easy but sloppy and incorrect API.

**Comment by droundy** [comment]
I want to qualify what I wrote about a path name based API.  Such an API can (I think) be entirely safe, if you force the use of the "fh" field when appropriate, e.g. by not providing paths on read/write, etc.  So far as I know this would make the API safe, since then when a file is opened, you'd have to create some sort of reference to it other than its path, so it could continue being read after it is deleted.

**Comment by wfraser** [comment]
Correct. Any filesystem that relies entirely on the `path` parameter in `read` or `write` is broken. The file handle parameter *has* to be used. The same if you use the C libfuse. In any function that takes a `fh` paremeter, the `path` is just a hint. For example, if it's got multiple hard-links, the given path is just the first one we saw for that inode; or if it's totally unlinked but still open, it's the last path seen for it, when really there isn't a valid path for it.

The only wrinkle here is with the `getattr` call, where rust-fuse doesn't currently pass the file handle, due to it not using a recent-enough version of the FUSE kernel ABI, but that's being worked on currently (see https://github.com/zargony/rust-fuse/issues/82). I know it prevents `tail -f` from tracking the file across renames, because it causes its `getattr` call to fail with `ENOENT` even though `tail` uses the file descriptor in its call; the file descriptor doesn't make it to FUSE yet.
